### PR TITLE
[API] Patchable Budgets

### PR DESCRIPTION
**Pull request by NobodysNightmare**
## OpenProject work package

https://community.openproject.org/work_packages/19141
## Dependencies

Requires merge of:
- [x] https://github.com/opf/openproject/pull/2708
- [x] https://github.com/opf/openproject/pull/2740
## TODO
- [x] document budget endpoints and links
- [x] add budget endpoints
- [x] link budget in WP
- [x] add budget to schema
- [x] add budget to payload (actually making it PATCH-able)
- [x] ~~adapt front end after renaming cost_object ("budget")~~
### Integration tests
- [ ] GET work package
- [x] POST work package form
- [x] PATCH work package
- [x] GET budget
- [x] GET budgets by project
## TODOs to maintain consistency
- [x] document changed behaviour of spent time on WP
- [x] reflect changed behaviour of spent time in schema

### CODE REVIEW

**Comment by NobodysNightmare** [comment]
Note to be safe:

I reduced the amount of properties available in the representer for cost objects. I don't think that the Angular client relied on anything beyond subject and id, but I could be wrong.

File: lib/open_project/costs/patches/work_package_patch.rb [hunk 1]
```diff
@@ -94,10 +94,14 @@ def update_cost_entries(work_packages, action)
...
   module InstanceMethods
+    def costs_enabled?
+      project && project.costs_enabled?
+    end
+
     def validate_cost_object
-      if cost_object && cost_object.changed?
-        unless (cost_object.blank? || project.cost_object_ids.include?(cost_object.id))
-          errors.add :cost_object, :invalid
+      if changed.include? 'cost_object_id'
```
Comment (ulferts) [inline lib/open_project/costs/patches/work_package_patch.rb]: `cost_object_id_changed?` is easier to read

**Comment by ulferts** [comment]
@NobodysNightmare the budget is no longer displayed, because of the renaming from cost_object to budget. You want to give it a try in the front end?

**Comment by ulferts** [comment]
@NobodysNightmare I tried unsuccessfully to patch the work package to have a new budget. And there is no integration test that let's me assume that this should be working. On the other hand the PR is named "Patchable Budgets"...

**Comment by NobodysNightmare** [comment]
I chose not to do an integrational test for `GET` work package.
Because we already have a spec for the representer and I don't see which extra benefit the inegration test would offer.

However, I will create the other integration tests. They seem useful.

**Comment by NobodysNightmare** [comment]
> You want to give it a try in the front end?

I did, but I am apparently too stupid...

~~@0xF013 this is your turn again, what do I have to do here? It is not `_embedded/costObject` anymore, but `_embedded/budget`. Can't be too hard, right?~~

edit: see below

**Comment by NobodysNightmare** [comment]
I changed the link back to `costObject`, because the waqy we generate the field name for errors in our form sucks and I am either forced to refactor there or simply stay with `costObject` for the time being...

Still able to change that in the future, right?

**Comment by NobodysNightmare** [comment]
Note that there is a new (small) dependency PR in the core.

**Pull request merged**